transaction/src/validation/transaction_validator.rs

```diff
@@ -307,39 +303,64 @@ impl TransactionValidator {
+            IntentActorProof::User(intent_signatures) => {
+                // TODO: split into static validation part and runtime validation part to support more signatures
+                if intent_signatures.len() > MAX_NUMBER_OF_INTENT_SIGNATURES {
+                    return Err(SignatureValidationError::TooManySignatures);
+                }
 
-        // verify intent signature
-        let intent_payload = transaction.signed_intent.intent.to_bytes();
-        let mut signers = HashSet::new();
-        for sig in &transaction.signed_intent.intent_signatures {
-            let public_key = recover(&intent_payload, sig)
-                .ok_or(SignatureValidationError::InvalidIntentSignature)?;
+                // verify intent signature
+                let mut signers = HashSet::new();
+                let intent_payload = transaction.signed_intent.intent.to_bytes();
+                for sig in intent_signatures {
+                    let public_key = recover(&intent_payload, sig)
+                        .ok_or(SignatureValidationError::InvalidIntentSignature)?;
 
-            if !verify(&intent_payload, &public_key, &sig.signature()) {
-                return Err(SignatureValidationError::InvalidIntentSignature);
-            }
+                    if !verify(&intent_payload, &public_key, &sig.signature()) {
+                        return Err(SignatureValidationError::InvalidIntentSignature);
+                    }
+
+                    if !signers.insert(public_key) {
+                        return Err(SignatureValidationError::DuplicateSigner);
+                    }
+                }
+
+                if transaction.signed_intent.intent.header.notary_as_signatory {
+                    signers.insert(transaction.signed_intent.intent.header.notary_public_key);
+                }
 
-            if !signers.insert(public_key) {
-                return Err(SignatureValidationError::DuplicateSigner);
+                // verify notary signature
+                let signed_intent_payload = transaction.signed_intent.to_bytes();
+                if !verify(
+                    &signed_intent_payload,
+                    &transaction.signed_intent.intent.header.notary_public_key,
+                    &transaction.notary_signature,
+                ) {
+                    return Err(SignatureValidationError::InvalidNotarySignature);
+                }
+
+                Ok(ValidatedTransactionActor::User(
+                    signers.into_iter().collect(),
+                ))
             }
-        }
 
-        // verify notary signature
-        let signed_intent_payload = transaction.signed_intent.to_bytes();
-        if !verify(
-            &signed_intent_payload,
-            &transaction.signed_intent.intent.header.notary_public_key,
-            &transaction.notary_signature,
-        ) {
-            return Err(SignatureValidationError::InvalidNotarySignature);
-        }
+            IntentActorProof::Supervisor => {
```

Comment:
What is the purpose of the supervisor? What does Supervisor mean?

To me, Supervisor suggests a higher-tier of user, and  suggests there's a third, even higher tier. I think I'd rather the term System :)

But yes - just to flag, I'm very concerned with the current impl that the TransactionValidator allows Supervisor signatures through unchallenged - because this means users could submit Supervisor transactions through the current node and have these work :)

IMO either:
* Supervisor transactions shouldn't use the TransactionValidator - in fact, they should go via an entirely different route, as they shouldn't take a NotarizedTransaction as input imo :)
* As a temporary workaround, TransactionValidator should have an option whether to allow/require Supervisor or User transactions - and default to only allowing User transactions.

Reply:
In fact, to double down, we should keep: `transaction.signed_intent.intent_signatures` and not change it to `intent_actor_proof` imo.

A NotarizedTransaction should always be a user transaction.

Reply:
`Supervisor` refers to the updates which the validatorSet can perform (e.g. roundUpdates). A higher tier would be `System` which is something like genesis where almost arbitrary updates can be performed (This would be the same mechanism used for forking updates).

> because this means users could submit Supervisor transactions through the current node and have these work... which is scary to me
Yes, this was going to be resolved at the node side checking for these Supervisor transactions but I think you have a good point that `TransactionValidator` may be the right place to put this. This is essentially like a `sudo` check.

I don't think these should go down an entirely different route. We've got a good auth model in the engine and we should just leverage that. There is definitely some difference between the two though and I think we can slowly move the enum split between `User` and `Supervisor` higher as gain more info on where the right place is. I'd prefer not to split from `NotarizedTransaction` right now though.